c++/src/kj/compat/tls.c++

```diff
@@ -405,31 +416,108 @@ private:
 // =======================================================================================
 // Implementations of ConnectionReceiver, NetworkAddress, and Network as wrappers adding TLS.
 
-class TlsConnectionReceiver final: public kj::ConnectionReceiver {
+template <typename T>
+class InputQueue {
+  // InputQueue is an async FIFO queue that has many sources but only one sink.
+  // TODO(pr) Should I find a header home for this somewhere?
+
 public:
-  TlsConnectionReceiver(TlsContext& tls, kj::Own<kj::ConnectionReceiver> inner)
-      : tls(tls), inner(kj::mv(inner)) {}
+  void pushRaw(T&& v) {
+    // Push an existing value onto the queue.
 
-  Promise<Own<AsyncIoStream>> accept() override {
-    return inner->accept().then([this](kj::Own<AsyncIoStream> stream) {
-      return kj::evalNow([&] { return tls.wrapServer(kj::mv(stream)); })
-          .catch_([this](kj::Exception&& e) {
-        KJ_LOG(ERROR, "error accepting tls connection", e);
-        return accept();
-      });
+    if (auto localFulfiller = kj::mv(fulfiller)) {
+      // Fast path: We have a consumer waiting.
+      localFulfiller->fulfill(kj::fwd<T>(v));
+      return;
+    }
+
+    // Slow path: We don't have a consumer, add it to the queue.
+    auto node = kj::heap<Node>(kj::fwd<T>(v));
+    if (head) {
+      Node* currentNode = head.get();
+      while (currentNode->next) {
+        // Run through the nodes until we don't have a next.
```

Comment:
FWIW it would be OK to use `std::deque` for this. Generally I think linked lists work better if the element type can have the `next` pointer embedded directly, avoiding the need to wrap it in a `Node` type that implies an extra allocation.

My one reservation with using `std::deque` is if we put this in a public header then that header has to include the std header, and we generally have a rule against including std headers in KJ headers for fear of people relying on the transitive includes, but I guess that's a pretty weak argument. (Also, using std container types always seems to lead to lots of frustration as they tend to get horribly confused by element types that are move-only or enforce transitive constness as KJ does. It's possible to work around, but annoying.)

Anyway up to you if you want to go that route or do the tail pointer thing.

Reply:
I was wondering if doing tail linking would be worthwhile. Pretty easy to implement that. 

I wouldn't mind writing OwnList if you think it's valuable, I'd probably TODO it here. (My trivial read is that I can generalize List to do both Own and pointer impls.)

Reply:
> My trivial read is that I can generalize List to do both Own and pointer impls.

Eh I doubt that, it's pretty fundamentally different. With non-owned elements, the expectation is that the elements will remove themselves from the list when destroyed. It's commonly used to represent a list of waiters waiting for some event, where each waiter can potentially be canceled. (If you generalize this queue to allow multiple waiters, you should probably use a List<T> there.)

Reply:
If I tried to use a List for the waiters, I'd need to have each waiter come in with a reference to its ListLink, right? This seems like it would be a more classical condVar+storage style of queue.

Reply:
Each call to `wait()` would use `newAdaptedPromise()` with an adapter defined privately in your queue class. The adapter would contain a ListLink.

Reply:
I think I understand the shape of what you're suggesting now. The primary benefit I'm seeing is that we'd combine the allocation of the PromiseFulfiller with the Node. Is that accurate?

Reply:
Right -- as opposed to using `OwnList<Own<PromiseFulfiller>>` or `std::list<Own<PromiseFulfiller>>`, that is.

In fact, I think `kj::newPromiseAndFulfiller` in itself does two or three allocations, whereas `newAdaptedPromise()` always does only one. So, even more savings.

And adapted promises can proactively remove themselves from the list on cancellation, rather than leaving behind orphaned fulfillers that have to be cleaned up later.